src/time.cr

```diff
@@ -159,19 +193,22 @@ struct Time
   # time.millisecond                   # => 789
   # ```
   def self.epoch_ms(milliseconds : Int) : self
-    new(UnixEpoch + milliseconds.to_i64 * Span::TicksPerMillisecond, Kind::Utc)
+    milliseconds = milliseconds.to_i64
+    seconds = UNIX_SECONDS + (milliseconds / 1_000)
+    nanoseconds = (milliseconds % 1000) * NANOSECONDS_PER_MILLISECOND
+    new(seconds: seconds, nanoseconds: nanoseconds.to_i, kind: Kind::Utc)
   end
 
   def clone
     self
   end
 
   def +(other : Span)
-    add_ticks other.ticks
+    add_nanoseconds other.total_nanoseconds
```

Comment:
What about using `other.total_seconds_i` and `other.nanoseconds` with manual nanosecond overflow checks but exact precision, instead of converting the span to a float, with division and modulo? Something like:

```crystal
def +(span : Span)
  s = @seconds + span.total_seconds_i
  ns = @nanoseconds + span.nanoseconds
  if ns > NANOSECONDS_PER_SECOND
    s += 1
    ns -= NANOSECONDS_PER_SECOND
  end
  new(seconds: s, nanoseconds: ns, kind: kind)
end
```

Reply:
Oh, yes. That code is old because I originally implemented this with Time::Span just having a nanoseconds instance var (so max was around 290 years, a bit limited)